include/bout/index_derivs.hxx

```diff
@@ -0,0 +1,826 @@
+    AUTO_TRACE();
+    using namespace std::placeholders;
+
+    // Now we want to get the actual field type out of the TypeContainer
+    // used to pass this around
+    using FieldType = typename FieldTypeContainer::type;
+
+    // Note whilst this should be known at compile time using this directly in the
+    // template parameters below causes problems for old versions of gcc/libstdc++
+    // (tested with 4.8.3) so we currently use a hacky workaround. Once we drop
+    // support for these versions the branching in the case statement below can be
+    // removed and we can use nGuard directly in the template statement.
+    const int nGuards = Method{}.meta.nGuards;
+
+    auto& derivativeRegister = DerivativeStore<FieldType>::getInstance();
+
+    switch (Method{}.meta.derivType) {
+    case (DERIV::Standard):
+    case (DERIV::StandardSecond):
+    case (DERIV::StandardFourth): {
+      if (nGuards == 1) {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template standard<Direction::value, Stagger::value, 1, FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      } else {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template standard<Direction::value, Stagger::value, 2, FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      }
+      break;
+    }
+    case (DERIV::Upwind):
+    case (DERIV::Flux): {
+      if (nGuards == 1) {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template upwindOrFlux<Direction::value, Stagger::value, 1,
+                                           FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3, _4);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
+            theFunc);
+      } else {
+        const auto theFunc = std::bind(
+            // Method to store in function
+            &Method::template upwindOrFlux<Direction::value, Stagger::value, 2,
+                                           FieldType>,
+            // Arguments -- first is hidden this of type-bound, others are placeholders
+            // for input field, output field, region
+            Method{}, _1, _2, _3, _4);
+        derivativeRegister.template registerDerivative<Direction, Stagger, Method>(
```

Comment:
Would it make sense for `registterDerivative` to take `Direction`, `Stagger`, and `Method` as arguments, rather than template parameters? Then the function could be called without the need to specify `template`

Reply:
There should be an overload already that would allow that but I think it would make the call slightly uglier potentially (if I remember correctly, which I might not).

Reply:
I think you can just change this overload:

```cpp
  template <typename Direction, typename Stagger, typename Method>
  void registerDerivative(standardFunc func) {
    AUTO_TRACE();
    registerDerivative(func, Method{}.meta.derivType, Direction{}.lookup(),
                       Stagger{}.lookup(), Method{}.meta.key);
  };
```

to take `Direction`, `Stagger` and `Method` by argument

Reply:
The `Direction`, `Stagger` and `Method` types here are wrapped up values/enums/types provided by the template combination code so we should be able to use type deduction with the templated routine to give 

~~~~
 /// Templated versions of the above registration routines.                                                                                                                                                       
  template <typename Direction, typename Stagger, typename Method>
  void registerDerivative(standardFunc func, Direction direction, Stagger stagger, Method method) {
    AUTO_TRACE();
    registerDerivative(func, method.meta.derivType, direction.lookup(),
                       stagger.lookup(), method.meta.key);
  };
~~~~

The call then looks like

~~~~
        derivativeRegister.registerDerivative(
            theFunc, Direction{}, Stagger{}, Method{});
~~~~